[wordpress__blocks] Add spacing and color defs to block supports

sunyatasattva opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [Block API: Block Supports](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/)

---

### Notes for this PR:

I have added some of the missing types to the `supports` section of a block, to be in line with what the docs currently say. However, I have two main questions about this:

1. **Should `__experimental` props be added to the types?** As of now, I have added them for two reasons: first of all, these props are actually there, so a user using them would be forced to manually edit the types or deal with frustrating errors; secondly, though **experimental**, they are not **internal**: they are actually mentioned in the official docs, that's why I believe they should be there, and we should update this package accordingly when they graduate or get dropped.
2. I am not sure which tests to add here, so I would appreciate some guidance: I have seen that the current tests do not really cover all of the possible `supports` properties, so I am not sure whether I should add tests here, and what kind of things should I test for this.

DangerBotOSS commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-Danger;
-->



Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## wordpress__blocks ([<kbd>unpkg</kbd>](https://unpkg.com/browse/wordpress__blocks@latest/))
was missing the following properties:
1. store
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 41d5bf2683c90310c4a85211e19b82534b7fb442
</p>

sirreal commented:
> * **Should `__experimental` props be added to the types?** As of now, I have added them for two reasons: first of all, these props are actually there, so a user using them would be forced to manually edit the types or deal with frustrating errors; secondly, though **experimental**, they are not **internal**: they are actually mentioned in the official docs, that's why I believe they should be there, and we should update this package accordingly when they graduate or get dropped.

I'd prefer to avoid adding anything experimental. They're guaranteed to change, either promoted to stable or removed entirely. 

> * I am not sure which tests to add here, so I would appreciate some guidance: I have seen that the current tests do not really cover all of the possible `supports` properties, so I am not sure whether I should add tests here, and what kind of things should I test for this.

Adding more supports to a declaration like this should be sufficient to test:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/41d5bf2683c90310c4a85211e19b82534b7fb442/types/wordpress__blocks/wordpress__blocks-tests.tsx#L396-L398

dmsnell commented:
to provide some additional context to what @sirreal mentioned, I concur that it's a good idea to avoid introducing the `__experimental` features into these types. we've been having discussion in Core talking about the emergence of these `__experimental` APIs and how they are leading to trouble because of how they encourage people to rely on them. that is, _all_ of the `__experimental` APIs are highly likely to quickly change or disappear; because of that, we should probably try and suggest that they don't exist in the types so that the types don't become another source of encouraging people to build their projects on code that will soon disappear.

I don't know if `__experimental` methods already exist in other DT packages for `@wordpress`, but even if they do, it's probably something that we should all kind of come back around to and think of removing.

sunyatasattva commented:
Hello folks!

Thanks for the answers and sorry for the late reply: I had some health issues that still need attending.

> to provide some additional context to what @sirreal mentioned, I concur that it's a good idea to avoid introducing the __experimental features into these types. we've been having discussion in Core talking about the emergence of these __experimental APIs and how they are leading to trouble because of how they encourage people to rely on them. 

I agree that, in general, the use of `__experimental` API by third-parties should be discouraged. However, I get mixed signals here: since they are documented in the official docs, it seems to me like it could be a strange mismatch between the text docs and types (which, in a way, are just another kind of documentation).

In any case, I'll go by this suggestion and remove them. I'll ping you once I have removed them and updated the tests.

dmsnell commented:
thanks @sunyatasattva, and it's truly a kind of bummer with `__experimental`, but also the docs. I think right now we're all trying to figure out a better way to do this.

please feel no rush to update; your health is more important. let me know if you need any help with this PR and I can try to provide it.

sunyatasattva commented:
Hey @dmsnell, @sirreal , I've updated my PR last week taking into account what we talked about. Would you mind taking a look and seeing if that's ready?

dmsnell commented on types/wordpress__blocks/index.d.ts:
```diff
@@ -19,6 +20,10 @@ declare module '@wordpress/data' {
 
+export type AxialDirections = 'horizontal' | 'vertical';
+
+export type CSSDirections = 'top' | 'right' | 'bottom' | 'left';
```
I think these are usually named in the singular, and arrays fill the plurality requirement. For example, `CSSDirection[]` is multiple CSS directions. A value of the given type can only inhabit one of the values so while it looks like we're discussing _all_ of them, we're actually saying, "A value of type `AxialDirections` can only be `horizontal` or `vertical`" which reads more awkwardly than "A value of the type `AxialDirection` can only be `horizontal` or `vertical`"

Don't want to block this PR because of that, so if you are willing consider renaming these two into the singular. That being said, I'll approve it in advance anyway.

dmsnell reviewed (approved):
Please consider the note I left. Thanks for everything!

sunyatasattva commented:
Ready to merge

dmsnell reviewed (approved):
Awesome, thanks again for all your work on this!

sunyatasattva commented:
Ready to merge
